### PR TITLE
osd - pg_interval_t::check_new_interval should not rely on pool.min_size to determine if the PG was active

### PR DESCRIPTION
**Pull request by guangyy**
If the pool's min_size is set improperly, during peering, pg_interval_t::check_new_interval
might wrongly determine the PG's state and cause the PG to stuck at down+peering forever

Fixes: #12162
Signed-off-by: Guang Yang <yguang@yahoo-inc.com>

### CODE REVIEW

**Comment by guangyy** [comment]
Hi @athanatos , would you please help taking a look? Thanks.

File: src/osd/PG.h [hunk 1]
```diff
@@ -279,6 +279,17 @@ class PG {
   bool is_ec_pg() const {
     return pool.info.ec_pool();
   }
+  // Convert the acting/up vector to a set of pg shards
+  void convert_to_pg_shards(const vector<int>& from, set<pg_shard_t>* to) const {
```
Comment (athanatos) [inline src/osd/PG.h]: This should be in osd_types.h.

**Comment by guangyy** [comment]
Hi @athanatos , One problem I am having to pass PGBackend::IsRecoverablePredicate to check_new_interval, is that it does not support forward declaration for PGBackend::IsRecoverablePredicate since it is an inner class.

How about passing PG * directly?

**Comment by guangyy** [comment]
@athanatos , updated according to our discussion in IRC.

File: src/osd/PG.h [hunk 2]
```diff
@@ -197,6 +197,9 @@ class PG {
   void update_snap_mapper_bits(uint32_t bits) {
     snap_mapper.update_bits(bits);
   }
```
Comment (athanatos) [inline src/osd/PG.h]: It's an annoying weakness of C++ that it's hard to indicate with the type system that you are giving ownership of the pointer to the caller.  Let's make it clear by adding

/// get_is_recoverable_predicate: caller owns returned pointer and must delete when done

File: src/osd/osd_types.cc [hunk 3]
```diff
@@ -2649,9 +2660,14 @@ bool pg_interval_t::check_new_interval(
       if (*p != CRUSH_ITEM_NONE)
 	++num_acting;
 
+    const pg_pool_t& old_pg_pool = lastmap->get_pools().find(pgid.pool())->second;
+    set<pg_shard_t> old_pg_shards;
```
Comment (athanatos) [inline src/osd/osd_types.cc]: old_acting_shards, not old_pg_shards

File: src/osd/osd_types.h [hunk 4]
```diff
@@ -880,6 +898,9 @@ struct pg_pool_t {
     return 0;
   }
 
+  /// converts the acting/up vector to a set of pg shards
+  void convert_to_pg_shards(const vector<int> &from, set<pg_shard_t>* to) const;
```
Comment (athanatos) [inline src/osd/osd_types.h]: Why is this in pg_pool_t?   Should probably be top level in osd_types.h

**Comment by guangyy** [comment]
@athanatos , rebased and fixed all the issue according to our discussions in IRC, would you mind adding this to test? Thanks.

**Pull request merged**